File: combinatorics/resource/ST/project_5/bai_toan_4/general_graph/am_to_amap.cpp
#include <iostream>
#include <map>
#include <vector>

using namespace std;

// Ham chuyen tu ma tran ke sang adjacency map
map<int, map<int, int>> amToAmap(const vector<vector<int>>& matrix) {
    int n = matrix.size();
    map<int, map<int, int>> amap;

    for (int u = 0; u < n; ++u) {
        for (int v = 0; v < n; ++v) {
            if (matrix[u][v] == 1) {
                amap[u][v] = 1; // Gan trong so la 1
            }
        }
    }

    return amap;
}

int main() {
    int n;
    cout << "Nhap so dinh: ";
    cin >> n;

    vector<vector<int>> matrix(n, vector<int>(n));
    cout << "Nhap ma tran ke (AM):\n";
    for (int i = 0; i < n; ++i)
        for (int j = 0; j < n; ++j)
            cin >> matrix[i][j];

    auto amap = amToAmap(matrix);

    cout << "Adjacency Map (AMap):\n";
    for (auto& [u, inner] : amap)
        for (auto& [v, w] : inner)
            cout << u << " -> " << v << ": " << w << "\n";

    return 0;
}

File: combinatorics/resource/ST/project_5/bai_toan_4/general_graph/am_to_amap.py
def am_to_amap(matrix):
    n = len(matrix)
    amap = {}

    for u in range(n):
        for v in range(n):
            if matrix[u][v] == 1:
                if u not in amap:
                    amap[u] = {}
                amap[u][v] = 1  # Gan trong so la 1 (do do thi khong co trong so)

    return amap

# Nhap tu ban phim
n = int(input("Nhap so dinh: "))
print("Nhap ma tran ke (AM):")
matrix = []
for _ in range(n):
    matrix.append(list(map(int, input().split())))

amap = am_to_amap(matrix)

print("Adjacency Map (AMap):")
for u in amap:
    for v in amap[u]:
        print(f"{u} -> {v}: {amap[u][v]}")